src/Components/Components/src/NavigationManager.cs

```diff
@@ -274,6 +281,128 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
         }
     }
 
+    /// <summary>
+    /// Notifies the registered <see cref="IHandleLocationChanging"/> handlers of the current location change.
+    /// </summary>
+    /// <param name="uri">The destination URI. This can be absolute, or relative to the base URI.</param>
+    /// <param name="isNavigationIntercepted">Whether this navigation was intercepted from a link.</param>
+    /// <param name="forceLoad">Whether the navigation is attempting to bypass client-side routing.</param>
+    /// <returns>A <see cref="ValueTask{TResult}"/> representing the completion of the operation.</returns>
+    public async ValueTask<bool> NotifyLocationChanging(string uri, bool isNavigationIntercepted, bool forceLoad)
+    {
+        _locationChangingCts?.Cancel();
+        _locationChangingCts = null;
+
+        if (_locationChangingHandlers.Count == 0)
+        {
+            return false;
+        }
+
+        _locationChangingCts = new();
+
+        var cancellationToken = _locationChangingCts.Token;
+        var context = new LocationChangingContext(uri, isNavigationIntercepted, forceLoad, cancellationToken);
+        var handlerCount = _locationChangingHandlers.Count;
+        var locationChangingHandlersCopy = ArrayPool<IHandleLocationChanging>.Shared.Rent(handlerCount);
+        _locationChangingHandlers.CopyTo(locationChangingHandlersCopy);
+
+        try
+        {
+            for (var i = 0; i < handlerCount; i++)
+            {
+                await locationChangingHandlersCopy[i].OnLocationChanging(context);
```

Comment:
This is interesting. We hadn't discussed whether the handlers would run in series or in parallel. (I don't mean on multiple threads BTW, I just mean with their async execution interleaved).

I had thought they would run in parallel but thinking about it now, I don't have any justification for why that should be.

Are there any pros/cons or important scenarios you're aware of that would push us to running them in series like this?

I guess one possible reason for running them in parallel would be if you had a handler that was something like this:

 1. Start saving the unsaved changes in the background
 2. If it takes too long or fails, bring up a dialog asking the user what they want to do

If you had more than one such handler, you might want them all to try doing their step 1 concurrently.

What do you think - are there any definite scenarios for requiring serial invocation?

Reply:
I guess the only case I can think of for serial invocation is if you have multiple handlers intentionally registered in an order such that the more expensive handlers run last, after the simpler handlers have confirmed that they think the navigation should continue. But I would imagine you could just put all that logic in a single handler anyway, so I'm not sure if this is a real scenario.

I think running the handlers in parallel sounds good to me.